staging/src/k8s.io/apiextensions-apiserver/pkg/apiserver/customresource_handler.go

```diff
@@ -468,7 +477,17 @@ func (r *crdHandler) createCustomResourceDefinition(obj interface{}) {
 	r.customStorageLock.Lock()
 	defer r.customStorageLock.Unlock()
 	// this could happen if the create event is merged from create-update events
-	r.removeStorage_locked(crd.UID)
+	tearDownFinished := r.removeStorage_locked(crd.UID)
+	if utilfeature.DefaultFeatureGate.Enabled(features.StorageVersionAPI) &&
+		utilfeature.DefaultFeatureGate.Enabled(features.APIServerIdentity) {
+		go func() {
+			// make sure no in-fight request to old storage exists
+			<-tearDownFinished
+			if err := r.updateStorageVersionFor(crd); err != nil {
+				utilruntime.HandleError(err)
```

Comment:
I'd like the update to block the handler chain on update failures, otherwise the storage migration won't be triggered when it should. Here I accept to not block because 1. the chance of storage version update failure is low 2. it's not ideal but crd authors can self-heal by sending another CRD-versioning update

Reply:
If we know that the storage version really changed, and if we had locks per CRD, we could be much more aggressive with blocking until this call returns.

Reply:
After some thoughts, I think we might not need a hard block for custom resources. For built-in resources, we require the apiserver to update storageversion before serving writes, because otherwise one apiserver can sneak in writing some data with a different storageversion, and die without updating storageversion. In that case the storage migrator won't be able to tell and won't trigger migration.

For custom resources, all apiserver replicas write the same storageversion update when a CRD changes its storageversion. It's unlikely for all apiservers to fail update storageversion (conflict won't cause that; restarted server will retry update; the only way is to kill all servers forever, which doesn't make sense).

> the storage migration won't be triggered when it should

It's totally fine for storage migration to kick in after the servers have served some write requests with new storage, as long as:
1. it can tell a migration is needed when encoding versions become consistent-- one apiserver succeeded to update storageversion and change the resource version is good enough
2. encoding versions eventually become consistent-- maybe we just need to allow more retries if the failure rate bothers us.

Reply:
@caesarxuchao convinced me that blocking new handler on storageversion update is necessary, and this PR doesn't have to get in 1.20 for StorageVersionAPI alpha. I will revisit locks per CRD later. Thanks for the review